README

 **
 * Magento
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@magentocommerce.com so we can send you a copy immediately.
 *
 * package    Netzarbeiter_CustomerActivation
 * copyright  Copyright (c) 2009 Vinai Kopp http://netzarbeiter.com/
 * license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 **


Magento Module: Netzarbeiter/CustomerActivation
Author: Vinai Kopp <vinai@netzarbeiter.com>


USAGE

This small extension makes it impossible for a customer to log in to magento,
until the account has been activated in the adminhtml interface.

Customers > Manage Customers > (select customer) > Account Information > Is activated

The Extension was designed to be used together with the Netzarbeiter_LoginCatalog extension
Netzarbeiter_LoginCatalog only shows products to customers that are logged in.
There also is the Extension Netzarbeiter_GroupsCatalog which may suit your needs better,
you have to evaluate yourself what is the best solution for your case.

You can disable the Extension on a global or website level using the options found under
System > Configuration > Customer > Customer Activation

You can also select email notifications for the admin (when a new customer registers) and
for the customer (wehn her account has been activated).

With the latest release this module now also works when customer email confirmation is turned
off.

Thanks to SeL for the french translation and the bugfix!
Thanks to SeL also for reporting the backend customer creation bug.
Thanks to Toon van Veelen for the dutch translation!
Thanks to Erik Hoeksma for the Greek translation!
Thanks to PaulE for the integration in the admin grid for mass activation of customers!
Thanks to Finn Snaterse for the inclusion in the Adminhtml Grid!

KNOWN BUGS:
- None! :D

TODO:
- Add customer activation as a Resource under Roles

If you have ideas for improvements or find bugs, please send them to vinai@netzarbeiter.com,
with Netzarbeiter_CustomerActivation as part of the subject line.



Changelog
Changelog for Netzarbeiter_CustomerActivation

0.5.5 2015-10-27
- Make mass action admin controller route compatible with SUPEE-6788 patch 

0.5.4 2014-11-19
- Bug fix: correctly evaluate admin_always_active setting in admin customer grid 

0.5.3 2014-09-13
- Activate all existing customers by default during installation
- Refactoring and improved tests to help keep module maintainable

0.5.2 2014-01-09
- Compatibility with older Magento versions without automatic VAT ID validation (I think 1.5 and older)

0.5.1 2013-12-20
- Remove dependency on PHP 5.3.2 or newer. Should work down to 5.2 again.

0.5.0 2013-11-24
- Hide post registration message to enter address for VAT validation for deactivated accounts.
- Correct handling if default activation status is true and also activate by group is true.
  In that case, the default activation status takes precedence since that is also how the
  backend config dependencies show the fields.

0.4.9 2013-11-15
- Implement workaround for versions of Magento where core/template::getDesignConfig
  is protected.

0.4.8 2013-11-12
- Fix de-association of selected default addresses on mass updates
- Only send notifications to customers whos status actually changed during a mass update 

0.4.7 2013-09-23
- Fix store variable in email templates
- Fix mass activation email notifications

0.4.6 2013-05-24
- Backward compatibility fix for Magento older then 1.7

0.4.5 2013-05-22
- Bugfix: enable usage default admin notification email template

0.4.4 2013-04-21
- Minor new feature: export to CSV or XML now contains the activation status column

0.4.3 2013-04-20
- New feature: require activation only for specific groups

0.4.2 2013-01-16
- Update composer dependency on stable magento-composer-installer

0.4.1 2012-12-05
- Bugfix: properly handle checkout registrations,
          support onepage and multishipping checkout.

0.4.0 2012-11-08
- Bugfix: only apply mass action to customer grid.

0.3.9 2012-09-25
- Change observer declaration to <type>model</type> to avoid strange bug a user was reporting

0.3.8 2012-08-07
- Don't reuse the customer model for mass actions - it messes up the addresses

0.3.7 2012-07-28
- Move customer grid rewrite to observer to avoid conflicts with other modules

0.3.6 2012-07-22
- Add configurable feature "Enable customer accounts by default"

0.3.5 2012-07-20
- Change session model for messages to core/session

0.3.4 2011-05-19
- Add _isAllowed() method to admin controller

0.3.3 2011-04-29
- Use better logic to figure out the customer store id
- Use config value variables in email templates (Thanks to Ryan Price once again!)

0.3.2 2011-04-28
- Added check of configuration values if emails should be sent (Thanks to Ryan Price for the patch!)

0.3.1 2011-03-23
- Added shipping and billing address models as email variables

0.3.0 2011-03-22
- Fix session logout handling for Magento 1.5 compatibility
- Added licence header comments to all files

0.2.9 2010-12-20
- Add japanese locale (thanks to Junya Sano for the translation!)

0.2.8 2010-12-20
- Drop Magento 1.3 Support
- Add customer_activated attribute to 1.4.2 (CE) / 1.9 (EE) customer_form_attribute table
- Add renderer for column to display null values as "not activated"

0.2.7 2010-03-22
- Add option to enable mass activation updates when module is disabled

0.2.6 2010-03-15
- Add it_IT locale

0.2.5 2010-03-11
- Fix account not activated message for magento 1.4
- Add customer_activated customer attribute to attribute_set so it is displayed in the admin interface in 1.4

0.2.4 2009-09-21
- Get rid of customer edit block override by setting the correct attribute source model

0.2.3 2009-08-10
- Make customer notification email template selectable on a store view configuraton level

0.2.2 2009-07-18
- Added new config option to keep module active in admin customer interface, even if it is disabled otherwise

0.2.1 2009-06-13
- Let through API calls unchecked

0.2.0 2009-06-13
- Added new Feature: send out notification emails to admin and customer

0.1.9 2009-06-13
- Added greek translation - thanks to Erik Hoeksma

0.1.8 2009-04-24
- Added dutch translation - thanks to Toon van Veelden!

0.1.7 2009-04-01
- Added Mass Action Controller and Amdin Customer Activation Status in the grid from PaulE - thanks for the contributon, man!

0.1.6 2009-01-13
- Declared stable

0.1.5 2008-12-30
- Bugfix in Admintml - create new customer form now works (again, thanks SeL for the bug report)

0.1.4 2008-11-13
- Bugfix in Adminhtml (thanks to SeL)
- French Translation (again, thanks to SeL)
- Bugfix, now also works when customer email activation is disabled

0.1.2 2008-08-29
- Use dropdown in Adminhtml interface instead of text field (now overrides Mage_Adminhtml_Block_Customer_Edit_Tab_Account)
- Translation of label and dropdown values in adminhtml interface

0.1.1 2008-08-28
- Bugfix for dependency problem in pear package

0.1.0 2008-08-25
- Initial release



app/code/community/Netzarbeiter/CustomerActivation/Model/Observer.php
<?php

/**
 * Magento
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@magentocommerce.com so we can send you a copy immediately.
 *
 * package    Netzarbeiter_CustomerActivation
 * copyright  Copyright (c) 2014 Vinai Kopp http://netzarbeiter.com/
 * license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */
class Netzarbeiter_CustomerActivation_Model_Observer
{
    const XML_PATH_ALWAYS_NOTIFY_ADMIN = 'customer/customeractivation/always_send_admin_email';

    /**
     * Fired on customer_login event
     * Check if the customer has been activated (via adminhtml)
     * If not, through login error
     *
     * @param Varien_Event_Observer $observer
     */
    public function customerLogin($observer)
    {
        $helper = Mage::helper('customeractivation');
        if (!$helper->isModuleActive()) {
            return;
        }

        if ($this->_isApiRequest()) {
            return;
        }

        $customer = $observer->getEvent()->getCustomer();
        $session = Mage::getSingleton('customer/session');

        if (!$customer->getCustomerActivated()) {
            /*
             * Fake the old logout() method without deleting the session and all messages
             */
            $session->setCustomer(Mage::getModel('customer/customer'))
                ->setId(null)
                ->setCustomerGroupId(Mage_Customer_Model_Group::NOT_LOGGED_IN_ID);

            if ($this->_checkRequestRoute('customer', 'account', 'createpost')) {
                /*
                 * If this is a regular registration, simply display message
                 */
                $message = $helper->__('Please wait for your account to be activated');

                $session->addSuccess($message);
            } else {
                /*
                 * All other types of login
                 */
                Mage::throwException($helper->__('This account is not activated.'));
            }
        }
    }

    /**
     * Flag new accounts as such
     *
     * @param Varien_Event_Observer $observer
     */
    public function customerSaveBefore($observer)
    {
        $customer = $observer->getEvent()->getCustomer();

        $helper = Mage::helper('customeractivation');
        $storeId = $helper->getCustomerStoreId($customer);

        if (!$helper->isModuleActive($storeId)) {
            return;
        }

        if (!$customer->getId()) {
            $customer->setCustomerActivationNewAccount(true);
            if ($this->shouldSetDefaultActivationStatus()) {
                $groupId = $customer->getGroupId();
                $defaultStatus = $helper->getDefaultActivationStatus($groupId, $storeId);
                $customer->setCustomerActivated($defaultStatus);

                if (!$defaultStatus) {
                    // Suppress the "enter your billing address for VAT validation" message.
                    // This setting will not be saved, its just for this request.
                    $helper = Mage::helper('customer/address');
                    if (method_exists($helper, 'isVatValidationEnabled')) {
                        if (is_callable(array($helper, 'isVatValidationEnabled'))) {
                            if (Mage::helper('customer/address')->isVatValidationEnabled($storeId)) {
                                Mage::app()->getStore($storeId)->setConfig(
                                    Mage_Customer_Helper_Address::XML_PATH_VAT_VALIDATION_ENABLED, false
                                );
                            }
                        }
                    }
                }
            }
        }
    }

    /**
     * Send out emails
     *
     * @param Varien_Event_Observer $observer
     */
    public function customerSaveAfter($observer)
    {
        /** @var Mage_Customer_Model_Customer $customer */
        $customer = $observer->getEvent()->getCustomer();

        $helper = Mage::helper('customeractivation');
        $storeId = $helper->getCustomerStoreId($customer);

        if (!$helper->isModuleActive($storeId)) {
            return;
        }

        $groupId = $customer->getGroupId();
        $defaultStatus = $helper->getDefaultActivationStatus($groupId, $storeId);

        try {
            if (Mage::app()->getStore()->isAdmin()) {
                if (!$customer->getOrigData('customer_activated') && $customer->getCustomerActivated()) {
                    // Send customer email only if it isn't a new account and it isn't activated by default
                    if (!($customer->getCustomerActivationNewAccount() && $defaultStatus)) {
                        $helper->sendCustomerNotificationEmail($customer);
                    }
                }
            } else {
                if ($customer->getCustomerActivationNewAccount()) {
                    // Only notify the admin if the default is deactivated or the "always notify" flag is configured
                    $alwaysNotify = Mage::getStoreConfig(self::XML_PATH_ALWAYS_NOTIFY_ADMIN, $storeId);
                    if (!$defaultStatus || $alwaysNotify) {
                        $helper->sendAdminNotificationEmail($customer);
                    }
                }
                $customer->setCustomerActivationNewAccount(false);
            }
        } catch (Exception $e) {
            Mage::throwException($e->getMessage());
        }
    }

    public function salesConvertQuoteAddressToOrder(Varien_Event_Observer $observer)
    {
        /** @var $address Mage_Sales_Model_Quote_Address */
        $address = $observer->getEvent()->getAddress();
        $this->_abortCheckoutRegistration($address->getQuote());
    }

    /**
     * Abort registration during checkout if default activation status is false.
     *
     * Should work with: onepage checkout, multishipping checkout and custom
     * checkout types, as long as they use the standard converter model
     * Mage_Sales_Model_Convert_Quote.
     *
     * Expected state after checkout:
     * - Customer saved
     * - No order placed
     * - Guest quote still contains items
     * - Customer quote contains no items
     * - Customer redirected to login page
     * - Customer sees message
     *
     * @param Mage_Sales_Model_Quote $quote
     */
    protected function _abortCheckoutRegistration(Mage_Sales_Model_Quote $quote)
    {
        $helper = Mage::helper('customeractivation');
        if (!$helper->isModuleActive($quote->getStoreId())) {
            return;
        }

        if ($this->_isApiRequest()) {
            return;
        }

        if (!Mage::getSingleton('customer/session')->isLoggedIn() && !$quote->getCustomerIsGuest()) {
            // Order is being created by non-activated customer
            $customer = $quote->getCustomer()->save();
            if (!$customer->getCustomerActivated()) {
                // Abort order placement
                // Exception handling can not be assumed to be useful

                // Todo: merge guest quote to customer quote and save customer quote, but don't log customer in

                // Add message
                $message = $helper->__(
                    'Please wait for your account to be activated, then log in and continue with the checkout'
                );
                Mage::getSingleton('core/session')->addSuccess($message);

                // Handle redirect to login page
                $targetUrl = Mage::getUrl('customer/account/login');
                $response = Mage::app()->getResponse();

                if (Mage::app()->getRequest()->isAjax()) {
                    // Assume one page checkout
                    $result = array('redirect' => $targetUrl);
                    $response->setBody(Mage::helper('core')->jsonEncode($result));
                } else {
                    if ($response->canSendHeaders(true)) {
                        // Assume multishipping checkout
                        $response->clearHeader('location')
                            ->setRedirect($targetUrl);
                    }
                }
                $response->sendResponse();
                /* ugly, but we need to stop the further order processing */
                exit();
            }
        }
    }

    /**
     * Return true if the request is made via the api
     *
     * @return boolean
     */
    protected function _isApiRequest()
    {
        return Mage::app()->getRequest()->getModuleName() === 'api';
    }

    /**
     * Check the current module, controller and action against the given values.
     *
     * @param string $module
     * @param string $controller
     * @param string $action
     * @return bool
     */
    protected function _checkRequestRoute($module, $controller, $action)
    {
        $req = Mage::app()->getRequest();
        if (strtolower($req->getModuleName()) == $module
            && strtolower($req->getControllerName()) == $controller
            && strtolower($req->getActionName()) == $action
        ) {
            return true;
        }
        return false;
    }

    /**
     * Check the current controller and action match the passed names
     *
     * @param string $controller
     * @param string $action
     * @return bool
     */
    protected function _checkControllerAction($controller, $action)
    {
        $req = Mage::app()->getRequest();
        return $this->_checkRequestRoute($req->getModuleName(), $controller, $action);
    }

    /**
     * Add customer activation option to the mass action block.
     *
     * This can't be done during the block abstract event
     *
     * @param Varien_Event_Observer $observer
     */
    public function adminhtmlBlockHtmlBefore(Varien_Event_Observer $observer)
    {
        // Check the grid is the customer grid
        if ($observer->getBlock()->getId() != 'customerGrid') {
            return;
        }

        // Check if there is a massaction block and if yes, add the massaction for customeractivation
        $massBlock = $observer->getBlock()->getMassactionBlock();
        if ($massBlock) {
            /** @var $helper Netzarbeiter_CustomerActivation_Helper_Data */
            $helper = Mage::helper('customeractivation');

            if (!$helper->isModuleActiveInAdmin()) {
                return;
            }

            $noEmail = Netzarbeiter_CustomerActivation_Helper_Data::STATUS_ACTIVATE_WITHOUT_EMAIL;
            $withEmail = Netzarbeiter_CustomerActivation_Helper_Data::STATUS_ACTIVATE_WITH_EMAIL;
            $deactivate = Netzarbeiter_CustomerActivation_Helper_Data::STATUS_DEACTIVATE;

            $massBlock->addItem(
                'customer_activated',
                array(
                    'label' => $helper->__('Customer Activated'),
                    'url' => Mage::getUrl('adminhtml/customerActivation/massActivation'),
                    'additional' => array(
                        'status' => array(
                            'name' => 'customer_activated',
                            'type' => 'select',
                            'class' => 'required-entry',
                            'label' => $helper->__('Customer Activated'),
                            'values' => array(
                                $noEmail => $helper->__('Yes (No Notification)'),
                                $withEmail => $helper->__('Yes (With Notification)'),
                                $deactivate => $helper->__('No')
                            )
                        )
                    )
                )
            );
        }
    }

    /**
     * Add the customer_activated attribute to the customer grid collection
     *
     * @param Varien_Event_Observer $observer
     */
    public function eavCollectionAbstractLoadBefore(Varien_Event_Observer $observer)
    {
        if (!Mage::helper('customeractivation')->isModuleActiveInAdmin()) {
            return;
        }

        if (Mage::app()->getRequest()->getControllerName() !== 'customer') {
            return;
        }

        /** @var $collection Mage_Customer_Model_Resource_Customer_Collection */
        $collection = $observer->getEvent()->getCollection();

        // Only add attribute to customer collections
        $customerTypeId = Mage::getSingleton('eav/config')->getEntityType('customer')->getId();
        $collectionTypeId = $collection->getEntity()->getTypeId();
        if ($customerTypeId == $collectionTypeId) {
            $collection->addAttributeToSelect('customer_activated');

        }
    }

    /**
     * Add customer_activated column to CSV and XML exports
     *
     * @param Varien_Event_Observer $observer
     */
    public function coreBlockAbstractPrepareLayoutAfter(Varien_Event_Observer $observer)
    {
        if (!Mage::helper('customeractivation')->isModuleActiveInAdmin()) {
            return;
        }

        if (Mage::app()->getRequest()->getControllerName() !== 'customer') {
            return;
        }

        $block = $observer->getBlock();
        if ($block->getType() === 'adminhtml/customer_grid') {
            // I don't think we need to limit applying the column by action
            //$action = Mage::app()->getRequest()->getActionName();
            //if (in_array($action, array('grid', 'index', 'exportCsv', 'exportXml'))) {

            $this->_addActivationStatusColumn($block);

            //}
        }
    }

    /**
     * Add the activation status column to the customer grid block.
     *
     * This is used from different events when displaying the block as well as
     * during exporting the grid to CSV or XML.
     *
     * @param Mage_Adminhtml_Block_Widget_Grid $block
     */
    protected function _addActivationStatusColumn(Mage_Adminhtml_Block_Widget_Grid $block)
    {
        /** @var $helper Netzarbeiter_CustomerActivation_Helper_Data */
        $helper = Mage::helper('customeractivation');

        // Add the attribute as a column to the grid
        $block->addColumnAfter(
            'customer_activated',
            array(
                'header' => $helper->__('Customer Activated'),
                'align' => 'center',
                'width' => '80px',
                'type' => 'options',
                'options' => array(
                    '0' => $helper->__('No'),
                    '1' => $helper->__('Yes')
                ),
                'default' => '0',
                'index' => 'customer_activated',
                'renderer' => 'customeractivation/adminhtml_widget_grid_column_renderer_boolean'
            ),
            'customer_since'
        );

        // Set the new columns order.. otherwise our column would be the last one
        $block->sortColumnsByOrder();
    }

    /**
     * Reportedly on Magento 1.6 customers are logged in automatically
     * by the lost password functionality (must be some customization actually).
     *
     * This observer method removes the customer id from the customer/session,
     * in effect causing a logout just in case.
     *
     * @param Varien_Event_Observer $observer
     */
    public function controllerActionPostdispatchCustomerAccountResetPasswordPost(Varien_Event_Observer $observer)
    {
        if (!Mage::helper('customeractivation')->isModuleActive()) {
            return;
        }
        if (version_compare(Mage::getVersion(), '1.7', '<')) {
            $session = Mage::getSingleton('customer/session');
            $customer = $session->getCustomer();
            if (!$customer->getCustomerActivated() && $session->isLoggedIn()) {
                $session->setCustomerId(null)->setId(null);
            }
        }

    }

    private function shouldSetDefaultActivationStatus()
    {
        return !$this->isAdminEditCustomerSaveAction() && !$this->_isApiRequest();
    }

    private function isAdminEditCustomerSaveAction()
    {
        return
            Mage::app()->getStore()->isAdmin() &&
            $this->_checkControllerAction('customer', 'save');
    }
}
